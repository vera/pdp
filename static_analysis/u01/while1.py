from pycparser import c_ast, parse_file

import sys

FILENAME = "while1.c"

class ExerciseChecker(c_ast.NodeVisitor):
	def __init__(self):
		self.number_of_for_loops = 0
		self.number_of_while_loops = 0
		super()

	def visit_For(self, node):
		self.number_of_for_loops += 1

	def visit_While(self, node):
		self.number_of_while_loops += 1

def check_exercise(filename):
	errors = []

	ast = parse_file(filename, use_cpp=True, cpp_path="gcc", cpp_args=["-nostdinc", "-E", r"-Ipycparser/utils/fake_libc_include"])

	v = ExerciseChecker()
	v.visit(ast)
	if v.number_of_for_loops != 0:
		errors.append("Das Programm sollte keine for-Schleifen enthalten.")
	if v.number_of_while_loops != 3:
		errors.append("Das Programm sollte genau drei while-Schleifen enthalten.")
		
	return errors

if __name__ == "__main__":
	errors = check_exercise(FILENAME)
	for error in errors:
		print("-", error)
	if len(errors) > 0:
		sys.exit(1)

