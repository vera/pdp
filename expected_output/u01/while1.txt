12, 10, 8, 6, 4, 2, 0
-1, -3, -5, -7, -9, -11, -13
0, 1, 4, 9, 16, 25, 36, 49, 64, 81
